debian/theme-manager.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2022 Himadri Sekhar Basu <hsb10@iitbbs.ac.in>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH theme-manager SECTION "May 29 2022"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
theme-manager \- program to do something
.SH SYNOPSIS
.B theme-manager
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B theme-manager
and
.B bar
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBtheme-manager\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR bar (1),
.BR baz (1).
.br
The programs are documented fully by
.IR "The Rise and Fall of a Fooish Bar" ,
available via the Info system.


debian/changelog
theme-manager (1.0.8) stable; urgency=medium

  * Update readme from version 1.0.6 to 1.0.7 (#22)
  * Create only one log file per user (#23)
  * Show logs on user request from menu (#24)
  * pythonize theme-manager (#25)
    - use pure python packaging methods
    - update .gitignore
    - update desktop files
    - Parse arguments from cmdline
    - Turn off debug mode by default, use '-v[erbose]'
      to enable it
    - Add options to parse arguments for help, indicator
      and version
    - read version from d/changelog
  * re-debianize to match python packaging (#26)
    - d/compat: deleted
    - d/control: update build and runtime depends
    - d/control: update standards
    - d/rules: use pybuild
    - d/copyright: update
    - d/install: remove 'etc'
    - d/postinst: update
    - d/postrm: add file
    - d/salsa-ci: add
    - d/theme-manager-doc.docs: add
    - d/theme-manager.1: add
  * update workflows (#27)
    - remove old source files
    - Update auto build (CI) workflow
    - Update publish deb packaging workflow
  * Minor fixes (#28)
    - src/main: fix wrong message for '-i'
    - src/gui: remove unused Lines and
      change postion of proctitle
  * Update readme with new manual install (#29)
    - fix logo
    - Update instructions for manual installation

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Sat, 04 Jun 2022 16:36:42 +0530

theme-manager (1.0.7) stable; urgency=medium

  * Reduce logging to file and more while debugging (#16)
    - Do not log current state and DE for every monitor cycle to file
    - Log current state and DE to file only when theme changes
  * Additional log message for user requested change (#17)
    - Insert separate log message when user requested
      new theme using Randomize button from main window
    - Insert separate log message when user requested
      new theme using Next button from indicator
  * Update readme with version 1.0.6 (#18)
  * Append random numbers after logfilename (#19)
    - For multiuser systems when a log file exists in
      /tmp, the app cannot recreate the logfile after
      logging into other user
  * fix logfile name (#20)
  * Add support to change plank theme (#21)

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Tue, 19 Apr 2022 08:58:21 +0530

theme-manager (1.0.6) stable; urgency=medium

  * Update readme with version 1.0.5
  * fix indicator; closes #10
  * enable logging for user info and debug; fixes #11 (#12)
    - DesktopTheme.py: log current and new theme info
    - common.py: log session, state and new theme info
    - theme-manager.py: log start of indicator
    - tm_daemon.py: log old and new states, and
      initation of daemon and indicator
  * Update debian/control
    - Update standards version
    - Add gir1.2-appindicator3-0.1 as dependency
    - Update description
  * Start indicator from main window
  * Update copyright year to 2022
  * Auto-change themes at user-defined interval (#13)
    - Update UI with notebook style
    - theme-manager.py: Add option to start indicator from
      main window
    - tm_daemon.py: Add two separate daemons to
      change theme on state change and at regular
      user-defined interval
  * Add gir1.2-appindicator3-0.1 as dependency (#14)
    - Update workflows
    - Update README
  * Add next theme option to indicator menu (#15)
    - tm_daemon.py: Add option to change theme from indicator
    - tm_daemon.py: Use system theme icon for indicator
    - theme-manager.py: fix icon in pop-up menu

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Fri, 08 Apr 2022 10:35:51 +0530

theme-manager (1.0.5) stable; urgency=medium

  * Update readme with version 1.0.4
  * fix for not updating to new state while running
  * add new and update old workflows
    - Update workflows to run on all branches
    - add new workflow to publish deb package from latest release
    - Fix typo in README
    - Update test script
  * Fix no icon error on ubuntu GNOME dash; fixes #7 (#8)
    - add generate_desktop_files
    - update desktop files using generate_desktop_files
    - Add theme-manager icon to ui file 
  * Add indicator to run continuously in background
    - Add an indicator and create a monitoring class
    - Monitor change of state of day every minute
    - Change theme whenever state changes within a minute
  * add tray icon
  * update run script to take arguments
  * Add startup desktop file
    - Update generate_desktop_files to add autostart file
    - generate new autostart file
    - update debian packaging

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Fri, 18 Mar 2022 21:20:21 +0530

theme-manager (1.0.4) stable; urgency=medium

  * Add gitignore
  * Add support for gnome/unity desktops
    - Now properly supports (Probably) GNOME, Ubuntu:GNOME, Unity
    - Print current DE to help debugging
  * new icon for theme-manager
  * fix for deprecated metacity for Gnome and unity desktop
  * Add code quality and CI workflows

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Sun, 27 Feb 2022 20:17:34 +0530

theme-manager (1.0.3) stable; urgency=medium

  * Add gitignore
  * Add support for gnome/unity desktops
    - Now properly supports (Probably) GNOME, Ubuntu:GNOME, Unity
    - Print current DE to help debugging

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Wed, 29 Dec 2021 12:57:11 +0530

theme-manager (1.0.2) stable; urgency=medium

  * Properly show current colour variant in statusbar
  * Show current themes as dictionary

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Mon, 13 Dec 2021 21:01:31 +0530

theme-manager (1.0.1) stable; urgency=medium

  * Add support for mate desktop
  * Support Gnome and Ubuntu:Gnome (#2)
  * Fully supports Mint-Y theme in cinnamon
  * Ready to go with cinnamon desktop
  * Update ui
  * Add support for gnome desktop

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Mon, 29 Nov 2021 22:26:37 +0530

theme-manager (1.0.0) stable; urgency=medium

  * Initial release.


usr/lib/theme-manager/theme-manager.py
#!/usr/bin/python3

# Copyright (C) 2021-2022 Himadri Sekhar Basu <hsb10@iitbbs.ac.in>
#
# This file is part of theme-manager.
#
# theme-manager is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# theme-manager is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with theme-manager. If not, see <http://www.gnu.org/licenses/>
# or write to the Free Software Foundation, Inc., 51 Franklin Street,
# Fifth Floor, Boston, MA 02110-1301, USA..
#
# Author: Himadri Sekhar Basu <hsb10@iitbbs.ac.in>
#

# import the necessary modules!
from cmath import log
import getpass
import gettext
import gi
import glob
import locale
import logging
import setproctitle
import string
import sys
from threading import Thread
import warnings
from random import choice

# Suppress GTK deprecation warnings
warnings.filterwarnings("ignore")

gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, Gio

from common import CONFIG_FILE, ThemeManager
from tm_daemon import AppIndicator
from DesktopTheme import desktop_theme
# from LoginTheme import login_theme

# i18n
APP = 'theme-manager'
LOCALE_DIR = "/usr/share/locale"
locale.bindtextdomain(APP, LOCALE_DIR)
gettext.bindtextdomain(APP, LOCALE_DIR)
gettext.textdomain(APP)
_ = gettext.gettext

setproctitle.setproctitle(APP)

UI_PATH = "/usr/share/"+APP+"/ui/"

## Setup logging
def create_logfile():
	logpath = '/tmp/'
	dlimitter = '_'
	username = getpass.getuser()
	random_code =  ''.join(choice(string.digits) for _ in range(4))
	if len(glob.glob(logpath+APP+dlimitter+username+'*')) ==0:
		logfile = logpath + APP + dlimitter + username + dlimitter + random_code + '.log'
	else:
		logfile = glob.glob(logpath+APP+dlimitter+username+'*')[0]
	
	return logfile

# Create logger
logger = logging.getLogger('Theme Manager')
# Set logging level
logger.setLevel(logging.DEBUG)

# Create StreamHandler which logs even debug messages
cHandler = logging.StreamHandler()
# Set level for StreamHandler
cHandler.setLevel(logging.DEBUG)

# create file handler which logs only info messages
# Set the log filename
logfile = create_logfile()
fHandler = logging.FileHandler(logfile)
# Set level for FileHandler
fHandler.setLevel(logging.INFO)

# create formatter and add it to the handlers
log_format = logging.Formatter('%(asctime)s %(name)s - %(levelname)s: %(message)s')
cHandler.setFormatter(log_format)
fHandler.setFormatter(log_format)

# add the handlers to the logger
logger.addHandler(cHandler)
logger.addHandler(fHandler)


class theme_manager(Gtk.Application):
	# Main initialization routine
	def __init__(self, application_id, flags):
		Gtk.Application.__init__(self, application_id=application_id, flags=flags)
		self.connect("activate", self.activate)
	
	def activate(self, application):
		windows = self.get_windows()
		if (len(windows) > 0):
			window = windows[0]
			window.present()
			window.show()
		else:
			window = ThemeManagerWindow(self)
			self.add_window(window.window)
			window.window.show()

class ThemeManagerWindow():
	
	def __init__(self, application):
		
		self.application = application
		self.settings = Gio.Settings(schema_id="org.x.theme-manager")
		self.manager = ThemeManager()
		self.destop_manager = desktop_theme()
		self.icon_theme = Gtk.IconTheme.get_default()
		
		# Set the Glade file
		gladefile = UI_PATH+"theme-manager.ui"
		self.builder = Gtk.Builder()
		self.builder.add_from_file(gladefile)
		self.window = self.builder.get_object("MainWindow")
		self.window.set_title(_("Theme Manager"))
		
		# Create variables to quickly access dynamic widgets
		self.statusbar = self.builder.get_object("status_bar")
		# input values
		self.color_variants = self.builder.get_object("colour_variants")
		self.systemtheme_variants = self.builder.get_object("system_theme_name")
		self.icontheme_variants = self.builder.get_object("Icon_theme_name")
		self.cursortheme_variants = self.builder.get_object("cursor_theme_name")
		self.user_interval_HH = self.builder.get_object("hour")
		self.user_interval_MM = self.builder.get_object("minute")
		self.user_interval_SS = self.builder.get_object("second")
		
		# Buttons
		self.randomize_button = self.builder.get_object("randomize_theme_button")
		self.save_button = self.builder.get_object("save_button")
		
		# Widget signals
		self.randomize_button.connect("clicked", self.on_random_button)
		self.save_button.connect("clicked", self.on_save_button)
		# self.quit_button.connect("clicked", self.on_quit)
		
		# Menubar
		accel_group = Gtk.AccelGroup()
		self.window.add_accel_group(accel_group)
		menu = self.builder.get_object("main_menu")
		# Add "Start Indicator" option in drop-down menu
		item = Gtk.ImageMenuItem()
		item.set_image(Gtk.Image.new_from_icon_name("theme-manager", Gtk.IconSize.MENU))
		item.set_label(_("Start Indicator"))
		item.connect("activate", self.start_indicator)
		key, mod = Gtk.accelerator_parse("<Control>I")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Add "Show Logs" option in drop-down menu
		item = Gtk.ImageMenuItem()
		item.set_image(Gtk.Image.new_from_icon_name("text-x-log", Gtk.IconSize.MENU))
		item.set_label(_("Show Logs"))
		item.connect("activate", self.show_logs)
		key, mod = Gtk.accelerator_parse("<Control>L")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Add "About" option in drop-down menu
		item = Gtk.ImageMenuItem()
		item.set_image(Gtk.Image.new_from_icon_name("help-about-symbolic", Gtk.IconSize.MENU))
		item.set_label(_("About"))
		item.connect("activate", self.open_about)
		key, mod = Gtk.accelerator_parse("F1")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Add "Quit" option in drop-down menu
		item = Gtk.ImageMenuItem(label=_("Quit"))
		image = Gtk.Image.new_from_icon_name("application-exit-symbolic", Gtk.IconSize.MENU)
		item.set_image(image)
		item.connect('activate', self.on_quit)
		key, mod = Gtk.accelerator_parse("<Control>Q")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		key, mod = Gtk.accelerator_parse("<Control>W")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Show all drop-down menu options
		menu.show_all()
		
		self.load_conf()
		self.state = self.manager.get_state_info()
		self.currenttheme = self.destop_manager.get_desktop_theme(self.state, self.manager.systemthemename, self.manager.colvariants)
		self.current_status()
	
	def load_conf(self):
		
		self.manager.load_config()
		self.color_variants.set_text(str(self.manager.colorvariants))
		self.systemtheme_variants.set_text(str(self.manager.systemthemename))
		self.icontheme_variants.set_text(str(self.manager.iconthemename))
		self.cursortheme_variants.set_text(str(self.manager.cursorthemename))
		self.user_interval_HH.set_value(self.manager.theme_interval_HH)
		self.user_interval_MM.set_value(self.manager.theme_interval_MM)
		self.user_interval_SS.set_value(self.manager.theme_interval_SS)
	
	def open_about(self, widget):
		dlg = Gtk.AboutDialog()
		dlg.set_transient_for(self.window)
		dlg.set_icon_name("theme-manager")
		dlg.set_logo_icon_name("theme-manager")
		dlg.set_title(_("About"))
		
		dlg.set_program_name(_("Theme Manager"))
		dlg.set_version("__DEB_VERSION__")
		dlg.set_comments(_("A Python3-based GUI application to change different colour variants of GTK, Icon, Cursor and other themes."))
		dlg.set_website("https://hsbasu.github.io/theme-manager")
		dlg.set_copyright("Copyright \xa9 2021-2022 Himadri Sekhar Basu")
		dlg.set_authors(["Himadri Sekhar Basu <https://hsbasu.github.io>"])
		dlg.set_documenters(["Himadri Sekhar Basu <https://hsbasu.github.io>"])
		try:
			h = open('/usr/share/common-licenses/GPL', encoding="utf-8")
			s = h.readlines()
			gpl = ""
			for line in s:
				gpl += line
			h.close()
			dlg.set_license(gpl)
		except Exception as e:
			print(e)
		
		def close(w, res):
			if res == Gtk.ResponseType.CANCEL or res == Gtk.ResponseType.DELETE_EVENT:
				w.destroy()
		dlg.connect("response", close)
		dlg.show()
	
	def on_quit(self, widget):
		self.application.quit()
	
	def on_random_button(self, widget):
		logger.info("User requested change using Randomize button.")
		self.state = self.manager.get_state_info()
		self.nexttheme = self.manager.prep_theme_variants(self.state)
		self.destop_manager.set_desktop_theme(self.state, self.nexttheme)
		self.currenttheme = self.destop_manager.get_desktop_theme(self.state, self.manager.systemthemename, self.manager.colvariants)
		self.current_status()
	
	def on_save_button(self, widget):
		"""Saves user configurations to config file.
		
		Saves user-defined configurations to config file.
		If the config file does not exist, it creates a new
		config file (~/.config/theme-manager/config.cfg)
		in user's home directory.
		"""
		Hr = str(self.user_interval_HH.get_value_as_int())
		Min = str(self.user_interval_MM.get_value_as_int())
		Sec = str(self.user_interval_SS.get_value_as_int())
		user_interval = Hr+':'+Min+':'+Sec
		self.manager.config['system-theme'] = {
			'color-variants': self.color_variants.get_text(),
			'system-theme-name': self.systemtheme_variants.get_text(),
			'icon-theme-name': self.icontheme_variants.get_text(),
			'cursor-theme-name': self.cursortheme_variants.get_text(),
			'theme-interval': user_interval
		}
		with open(CONFIG_FILE, 'w') as f:
				self.manager.config.write(f)
		
		self.load_conf()
	
	def current_status(self):
		'''
		Show current theme info in status bar.
		'''
		status = "DE: %s, \tState: %s, \tVariant: %s, \tLast Updated: %s, \tThemes: %s" % (self.state['DE'], self.state['State'], self.currenttheme["Variant"], self.currenttheme["Last Updated"], self.currenttheme["Themes"])
		
		context_id = self.statusbar.get_context_id("status")
		self.statusbar.push(context_id, status)
	
	def start_indicator(self, widget):
		logger.info("Initiaing Theme Manager Indicator from main window.")
		indicatordaemon = Thread(target=AppIndicator())
		indicatordaemon.setDaemon(True)
		indicatordaemon.start()
	
	def show_logs(self, widget):
		try:
			h = open(logfile, encoding="utf-8")
			s = h.readlines()
			logs = ""
			for line in s:
				logs += line
			h.close()
		except Exception as e:
			print(e)
		
		logger_ui = UI_PATH+"logger.ui"
		log_builder = Gtk.Builder()
		log_builder.add_from_file(logger_ui)
		logger_dlg = log_builder.get_object("logger_dlg")
		logger_dlg.set_transient_for(self.window)
		logger_dlg.set_title(_("Theme Manager Logs"))
		logger_dlg.add_buttons(Gtk.STOCK_CLOSE, Gtk.ResponseType.CLOSE)
		logview = log_builder.get_object("log_view")
		logview.set_editable(False)
		logview.set_wrap_mode(Gtk.WrapMode.WORD)
		logview.get_buffer().set_text(logs)
		
		def close(w, res):
			if res == Gtk.ResponseType.CLOSE or res == Gtk.ResponseType.DELETE_EVENT:
				w.destroy()
		logger_dlg.connect("response", close)
		logger_dlg.show()
		

if __name__ == "__main__":
	try:
		if sys.argv[1] == '--indicator':
			# initiaing app indicator
			logger.info("Initiaing Theme Manager Indicator.")
			AppIndicator()
	except:
		application = theme_manager("org.x.theme-manager", Gio.ApplicationFlags.FLAGS_NONE)
		application.run()
		